### PR TITLE
Some fixes for anonymous uberclone and additional options.

### PR DESCRIPTION
**Pull request by crummel**
- Fix clone operation to work without a real PAT.
  - Instead of hitting the GitHub API to get the dependency graph, clone the repo then get the dependencies locally.
  - Remove unneeded BAR dependency.
- Add two options for limiting the uberclone: depth limit and ignored repos.
- Add one option for relocating `.git` directories (to simulate the way submodules work in regular git).

### CODE REVIEW

**Comment by crummel** [comment]
I'm working on some perf improvements for this now as well if you'd like to wait and review all at once.

**Comment by crummel** [comment]
@mmitche @jcagme This is ready for review now.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using CSharpx;
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Does this affect source build-ability of this?

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 2]
```diff
@@ -28,25 +31,134 @@ public override async Task<int> ExecuteAsync()
...
                     IEnumerable<DependencyDetail>  rootDependencies = await local.GetDependenciesAsync();
-
-                    foreach (DependencyDetail dep in rootDependencies)
-                    {
-                        Logger.LogInformation($"Found local dependency {dep.RepoUri}@{dep.Commit}.  Starting deep clone...");
-                        await CloneRemoteRepoAndDependencies(remoteFactory, _options.ReposFolder, dep.RepoUri, dep.Commit, _options.IncludeToolset, Logger);
-                    }
+                    IEnumerable<StrippedDependency> stripped = rootDependencies.Select(d => new StrippedDependency(d));
+                    stripped.ForEach((s) => accumulatedDependencies.Add(s));
+                    stripped.ForEach((s) => seenDependencies.Add(s));
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: NIT: this is enumerating twice. Add to both lists in the same ForEach.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 3]
```diff
@@ -56,56 +168,139 @@ public override async Task<int> ExecuteAsync()
...
+                File.WriteAllText(Path.Combine(repoPath, ".git"), GetGitDirRedirectString(masterRepoGitDirPath));
+                log.LogInformation($"Checking out {commit} in {repoPath}");
+                local = new Local(log, repoPath);
+                local.Checkout(commit, true);
+            }
+
+            return local;
+        }
+
+        private async static Task HandleMasterCopy(RemoteFactory remoteFactory, string repoUrl, string masterGitRepoPath, string masterRepoGitDirPath, ILogger log)
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: You may want to extract some methods out of this. The method is large and somewhat hard to follow with all the if/else statements.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 4]
```diff
@@ -28,25 +31,134 @@ public override async Task<int> ExecuteAsync()
...
+                        Logger.LogInformation($"Reached clone depth limit, aborting with {accumulatedDependencies.Count} dependencies remaining");
+                        foreach (StrippedDependency d in accumulatedDependencies)
+                        {
+                            Logger.LogDebug($"Abandoning dependency {d.RepoUri}@{d.Commit}");
+                        }
+                        break;
+                    }
+                    else
+                    {
+                        _options.CloneDepth--;
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Do expect that this will converge in a typical graph? I don't think it will at times.  For instance, if there is a cycle in the graph (DotNet-Trusted and core-setup have a cycle today)

**Comment by mmitche** [comment]
>     internal class CloneOperation : Operation

Can you add some high level information about the algorithm here? The redirection of git dirs is not entirely intuitive when looking at the code.

---
Refers to: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs:18 in 56c66f9. [](commit_id = 56c66f9d68855010e40e963c4c2ada57ddaca036, deletion_comment = False)

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs (hunk 1 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Removed.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs (hunk 2 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Fixed.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 5]
```diff
@@ -28,25 +31,134 @@ public override async Task<int> ExecuteAsync()
...
+                                {
+                                    Logger.LogDebug($"Skipping self-dependency in {repo.RepoUri} ({repo.Commit} => {d.Commit})");
+                                }
+                                else if (_options.IgnoredRepos.Any(r => r.Equals(d.RepoUri, StringComparison.OrdinalIgnoreCase)))
+                                {
+                                    Logger.LogDebug($"Skipping ignored repo {d.RepoUri} (at {d.Commit})");
+                                }
+                                else if (string.IsNullOrWhiteSpace(d.Commit))
+                                {
+                                    Logger.LogWarning($"Skipping dependency from {repo.RepoUri}@{repo.Commit} to {d.RepoUri}: Missing commit.");
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: >LogWarning [](start = 43, length = 10)

Although probably this won't run as part of a build, most build definitions would treat warnings as errors failing the build.

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 6]
```diff
@@ -257,9 +257,172 @@ public Task<GitDiff> GitDiffAsync(string repoUri, string baseVersion, string tar
...
+                catch (Exception exc)
+                {
+                    throw new Exception($"Something went wrong when checking out {commit} in {repoDir}", exc);
+                }
+            }
+        }
+
+        private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger log)
+        {
+            log.LogDebug($"Beginning clean of {repo.Info.WorkingDirectory} and {repo.Submodules.Count()} submodules");
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >log.LogDebug [](start = 12, length = 12)

log.BeginScope

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 7]
```diff
@@ -257,9 +257,172 @@ public Task<GitDiff> GitDiffAsync(string repoUri, string baseVersion, string tar
...
+
+            foreach (LibGit2Sharp.Submodule sub in repo.Submodules)
+            {
+                string normalizedSubPath = sub.Path.Replace('\\', Path.DirectorySeparatorChar).Replace('/', Path.DirectorySeparatorChar);
+                string subRepoPath = Path.Combine(repo.Info.WorkingDirectory, normalizedSubPath);
+                string subRepoGitFilePath = Path.Combine(subRepoPath, ".git");
+                if (!File.Exists(subRepoGitFilePath))
+                {
+                    log.LogDebug($"Submodule {sub.Name} in {subRepoPath} does not appear to be initialized, attempting to initialize now.");
+                    // hasn't been initialized yet, can happen when differest hashes have new or moved submodules
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >differest [](start = 68, length = 9)

s/differst/different

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs [hunk 8]
```diff
@@ -257,9 +257,172 @@ public Task<GitDiff> GitDiffAsync(string repoUri, string baseVersion, string tar
...
+                    repo.Submodules.Update(sub.Name, new LibGit2Sharp.SubmoduleUpdateOptions { Init = true });
+                }
+
+                log.LogDebug($"Beginning clean of submodule {sub.Name}");
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: >log.LogDebug [](start = 16, length = 12)

log.LogDebug

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs [hunk 9]
```diff
@@ -161,31 +175,101 @@ protected void Clone(string repoUri, string commit, string targetDirectory, ILog
...
                 }
                 catch (Exception exc)
                 {
-                    throw new Exception($"Something went wrong when cloning repo {repoUri} at {commit} into {targetDirectory}", exc);
+                    throw new Exception($"Something went wrong when cloning repo {repoUri} at {commit ?? "<default branch>"} into {targetDirectory}", exc);
+                }
+            }
+        }
+
+        private static void CheckoutSubmodules(LibGit2Sharp.Repository repo, LibGit2Sharp.CloneOptions submoduleCloneOptions, string gitDirParentPath, ILogger log)
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs]: >private static  [](start = 8, length = 15)

Nit: maybe insert some blank lines in between

**Comment by jcagme** [comment]
Given that the used algorithms are a bit complex we should add some test cases using permutations that would generate graphs with different depths, graphs with cycles, etc.

**Comment by crummel** [comment]
>     internal class CloneOperation : Operation

Added.

---
In reply to: [482721983](https://github.com/dotnet/arcade-services/pull/326#issuecomment-482721983) [](ancestors = 482721983)

---
Refers to: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs:18 in 56c66f9. [](commit_id = 56c66f9d68855010e40e963c4c2ada57ddaca036, deletion_comment = False)

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs (hunk 5 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Honestly this seems like something that should be fixed if it comes up - it's probably a subscription that wasn't set up properly, right?  Or is there a valid use case were we won't have a hash for a Version.Details.xml dependency?

---
In reply to: [275515042](https://github.com/dotnet/arcade-services/pull/326#discussion_r275515042) [](ancestors = 275515042)

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs (hunk 4 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Think I have this handled right now, I'm testing it out.

---
In reply to: [275061860](https://github.com/dotnet/arcade-services/pull/326#discussion_r275061860) [](ancestors = 275061860)

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs (hunk 3 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Refactored this a bit.

---
In reply to: [275044780](https://github.com/dotnet/arcade-services/pull/326#discussion_r275044780) [](ancestors = 275044780)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 7 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Fixed.

---
In reply to: [275518129](https://github.com/dotnet/arcade-services/pull/326#discussion_r275518129) [](ancestors = 275518129)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 8 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Added.

---
In reply to: [275518244](https://github.com/dotnet/arcade-services/pull/326#discussion_r275518244) [](ancestors = 275518244)

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs (hunk 9 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs]: Spaced out a bit.

---
In reply to: [275519407](https://github.com/dotnet/arcade-services/pull/326#discussion_r275519407) [](ancestors = 275519407)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 8 above)
Comment (jcagme) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Sweet! I pasted the incorrect piece of code but you still got the idea!

---
In reply to: [275572174](https://github.com/dotnet/arcade-services/pull/326#discussion_r275572174) [](ancestors = 275572174,275518244)

File: src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs (hunk 6 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs]: Changed as suggested.

---
In reply to: [275517872](https://github.com/dotnet/arcade-services/pull/326#discussion_r275517872) [](ancestors = 275517872)

**Comment by crummel** [comment]
It doesn't really seem like we're set up for testing with mocks and I don't think we want to do this with actual clones - I might be missing something though, what do you think?  I could add an issue for improving testability.

---
In reply to: [483395000](https://github.com/dotnet/arcade-services/pull/326#issuecomment-483395000) [](ancestors = 483395000)

**Comment by crummel** [comment]
I think I've addressed everything now and rebased on latest master.  @jcagme - had a question about the testing you suggested.

**Comment by mmitche** [comment]
The testing I was talking about was under the src\Maestro\tests\Scenarios.  They use darc and Maestro to test a lot of E2E scenarios.  I'm fine with an actual clone there.  I'd pick a known good hash (e.g. preview5) that you expect to work and do a darc clone test of that.  Those will get run on rollout (when we make changes to darc or Maestro)

**Comment by crummel** [comment]
This test takes about 7.5 minutes.  Let me know if that's too long and I can reduce the depths, a test with depth 1 and 2 takes about 3 minutes.

**Comment by crummel** [comment]
Rebased again on latest.

**Comment by crummel** [comment]
@mmitche does this scenario test look okay?

**Comment by mmitche** [comment]
@crummel Looks great.  Thanks
